homeassistant/helpers/script.py

```diff
@@ -390,47 +423,108 @@ def async_script_wait(entity_id, from_s, to_s):
         finally:
             unsub()
 
+    async def _async_call_service_step(self):
+        """Call the service specified in the action."""
+        domain, service, service_data = self._prep_call_service_step()
+        blocking = service_data.get(ATTR_BLOCKING, True)
+
+        # If this might start a fully blocking script call then disable the call
+        # timeout. Otherwise use the normal service call limit.
+        if domain == "script" and (
+            service == SERVICE_TURN_ON and blocking or service != SERVICE_TURN_OFF
+        ):
+            limit = None
+        else:
+            limit = SERVICE_CALL_LIMIT
+
+        coro = self._hass.services.async_call(
+            domain,
+            service,
+            service_data,
+            blocking=True,
+            context=self._context,
+            limit=limit,
+        )
 
-class _BackgroundScriptRun(_ScriptRun):
-    """Manage background Script sequence run."""
+        if limit is None:
+            # No call limit (i.e., potentially starting a fully blocking script call) so
+            # watch for a stop request and, if received, propagate it to the called
+            # scripts by turning them off.
+            task = self._hass.async_create_task(coro)
+            done, pending = await asyncio.wait(
+                {self._stop.wait(), task}, return_when=asyncio.FIRST_COMPLETED,
+            )
+            for pending_task in pending:
+                if pending_task == task:
+                    # The scripts did not all finish. Stop them by turning them off.
+                    # Do not cancel them now. Stopping them should cause them to finish
+                    # gracefully.
+                    if service in (SERVICE_TOGGLE, SERVICE_TURN_ON):
+                        service_data = {
+                            key: value
+                            for key, value in service_data.items()
+                            if key in cv.ENTITY_SERVICE_FIELDS
+                        }
+                    else:
+                        service_data = {ATTR_ENTITY_ID: f"script.{service}"}
+                    await self._hass.services.async_call(
+                        "script", SERVICE_TURN_OFF, service_data, blocking=True
+                    )
+                else:
+                    pending_task.cancel()
+            # Propagate any exceptions that might have happened.
+            for done_task in done:
+                done_task.result()
+        else:
+            # No call limit, so just wait for it to finish.
+            await coro
 
-    async def async_run(self) -> None:
-        """Run script."""
-        self._hass.async_create_task(self._async_run(False))
 
+class _QueuedScriptRun(_ScriptRun):
+    """Manage queued Script sequence run."""
 
-class _BlockingScriptRun(_ScriptRun):
-    """Manage blocking Script sequence run."""
+    async def _async_run(self):
+        # pylint: disable=protected-access
+        _, pending = await asyncio.wait(
```

Comment:
Please add a comment to explain what this logic does for future readers.

Reply:
Will do.